### PR TITLE
refactor: Replace `std::optional<bilingual_str>` with `util::Result`

### PR DESCRIPTION
**Pull request by ryanofsky**
Replace uses of `std::optional<bilingual_str>` with `util::Result` as suggested https://github.com/bitcoin/bitcoin/pull/25648#discussion_r936311768, https://github.com/bitcoin/bitcoin/pull/27632#discussion_r1192007516, https://github.com/bitcoin/bitcoin/pull/27632#discussion_r1194858242, https://github.com/bitcoin/bitcoin/pull/24313#discussion_r1204047087

### CODE REVIEW

**Comment by ryanofsky** [comment]
Updated 9d3918d47daf7b118507ea6db0dcef69cedf7005 -> ebc493e94bbc43b6611b7211b77c12c453171ce9 ([`pr/bresult-opt.1`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-opt.1) -> [`pr/bresult-opt.2`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-opt.2), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/bresult-opt.1..pr/bresult-opt.2)) to fix CI compile error https://cirrus-ci.com/task/6195420723412992?logs=ci#L2419

**Comment by ryanofsky** [comment]
Thanks, added std::move as suggested.

Updated b2bcba068e2c5542ab157e62c82bb868f678beeb -> 8aa8f73adce72e1ae855b43413c1f65504423cb7 ([`pr/bresult-opt.10`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-opt.10) -> [`pr/bresult-opt.11`](https://github.com/ryanofsky/bitcoin/commits/pr/bresult-opt.11), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/bresult-opt.10..pr/bresult-opt.11)) with workaround for CI errors https://cirrus-ci.com/task/6456067106799616?logs=ci#L1528 https://cirrus-ci.com/task/5048692223246336?logs=ci#L1240 https://cirrus-ci.com/task/6174592130088960?logs=ci#L1328

```c++
addrdb.cpp:213:12: error: call to implicitly-deleted copy constructor of 'util::Result<std::__1::unique_ptr<AddrMan, std::__1::default_delete<AddrMan> > >::T' (aka 'std::__1::unique_ptr<AddrMan, std::__1::default_delete<AddrMan> >')
    return addrman;
           ^~~~~~~
/usr/lib/llvm-10/bin/../include/c++/v1/memory:2513:3: note: copy constructor is implicitly deleted because 'unique_ptr<AddrMan, std::__1::default_delete<AddrMan> >' has a user-declared move constructor
  unique_ptr(unique_ptr&& __u) _NOEXCEPT
  ^
./util/result.h:47:14: note: passing argument to parameter 'obj' here
    Result(T obj) : m_variant{std::in_place_index_t<1>{}, std::move(obj)} {}
             ^
1 error generated.
```

**Review by sedited — APPROVED** [review]
ACK 8aa8f73adce72e1ae855b43413c1f65504423cb7

I think `[[nodiscard]` qualifiers for all the `util::Result` return types would be nice.

File: src/util/result.h [hunk 1]
```diff
@@ -31,16 +31,19 @@ struct Error {
...
 private:
+    using T = std::conditional_t<std::is_same_v<M, void>, std::monostate, M>;
+
     std::variant<bilingual_str, T> m_variant;
 
     template <typename FT>
     friend bilingual_str ErrorString(const Result<FT>& result);
 
 public:
+    Result() : m_variant{std::in_place_index_t<1>{}, std::monostate{}} {}  // constructor for void
```
Comment (hebasto) [inline src/util/result.h:46]: 5f49cb1bc8e6ba0671c21bf6292d2d3de43fd001

nit: Add `#include <utility>`?

**Review by hebasto — APPROVED** [review]
ACK 8aa8f73adce72e1ae855b43413c1f65504423cb7, I have reviewed the code and it looks OK.

**Comment by hebasto** [comment]
> Looks like there is a clang bug. Could be fixed by either bumping it again, see #27682, but I am not sure how to do that for macOS. Or by replacing `return addrman;` with `return {std::move(addrman)};`, or something like that (temporarily).

Although GCC complains:
```
$ ./configure CONFIG_SITE=$PWD/depends/arm-linux-gnueabihf/share/config.site --enable-suppress-external-warnings CXXFLAGS="-Wextra -Wno-psabi"
$ make > /dev/null 
addrdb.cpp: In function ‘util::Result<std::unique_ptr<AddrMan> > LoadAddrman(const NetGroupManager&, const ArgsManager&)’:
addrdb.cpp:213:21: warning: redundant move in return statement [-Wredundant-move]
  213 |     return std::move(addrman); // std::move should be unneccessary but is temporarily needed to work around clang bug (https://github.com/bitcoin/bitcoin/pull/25977#issuecomment-1561270092)
      |            ~~~~~~~~~^~~~~~~~~
addrdb.cpp:213:21: note: remove ‘std::move’ call
```

**Pull request merged**